### PR TITLE
DEV: Convert sidebar integration tests to gjs

### PR DESCRIPTION
**Pull request by IgnaceMaes**
A test run for https://github.com/IgnaceMaes/ember-codemod-template-tag.

### CODE REVIEW

File: app/assets/javascripts/discourse/tests/integration/components/sidebar/section-test.gjs [hunk 1]
```diff
@@ -26,16 +29,21 @@ module("Integration | Component | sidebar | section", function (hooks) {
...
         @headerLinkTitle="some title"
         @headerActionsIcon="plus"
         @headerActions={{this.headerActions}}
-        @displaySection={{this.displaySection}}/>`;
-
-    this.displaySection = false;
+        @displaySection={{state.displaySection}}
+      />
+    </template>;
     this.headerActions = [];
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/tests/integration/components/sidebar/section-test.gjs:47]: Am I right in saying that we can't use `this.` at all in test `<template>`? If so, seems like `headerActions` should also be moved into the `state`? (even though it doesn't actually change)

File: app/assets/javascripts/discourse/tests/integration/components/sidebar/section-test.gjs [hunk 2]
```diff
@@ -44,19 +52,22 @@ module("Integration | Component | sidebar | section", function (hooks) {
...
   test("can expand and collapse content when section is collapsible", async function (assert) {
-    const template = hbs`
-      <Sidebar::Section
+    const template = <template>
+      <Section
         @sectionName="test"
         @headerLinkText="test header"
         @headerLinkTitle="some title"
         @headerActionsIcon="plus"
         @headerActions={{this.headerActions}}
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/tests/integration/components/sidebar/section-test.gjs:67]: I think this needs to be moved into `state`?

File: app/assets/javascripts/discourse/tests/integration/components/sidebar/section-test.gjs [hunk 3]
```diff
@@ -1,20 +1,23 @@
...
-    const template = hbs`
-      <Sidebar::Section
+    const template = <template>
+      <Section
         @sectionName="test"
         @headerLinkText="test header"
         @headerLinkTitle="some title"
         @headerActionsIcon="plus"
-        @headerActions={{this.headerActions}} />`;
+        @headerActions={{this.headerActions}}
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/tests/integration/components/sidebar/section-test.gjs:18]: Another one for `state`?

**Comment by chancancode** [comment]
I posted this on Discord, but just so it doesn't get lost:

I think we can transform the `const template = hbs` fully inline without introducing extra variables, modifying modifying the `this.*` reference in the template and the downstream `this.set` calls (which can be difficult to get exactly right in a codemod):

```js
const template = (() => {
  let set = (key) => {
    thrown new Error(`cannot set "${key}" before render`);
  };

  this.set = (key, value) => { set(key, value) };

  return class extends Component {
    constructor(...args) {
      super(...args);
      set = (key, value) => { Reflect.set(this, key, value) };
    }

    @tracked foo;

    <template>
      {{this.foo}}
    </template>
  }
})();

await render(template);
```

you can also extract the boilderplate to an imported helper, so you are just left with:

```js
const template = TestTemplate(this, class extends Component {
  @tracked foo;

  <template>
    {{this.foo}}
  </template>
})();

await render(template);
```

```
function TestTemplate(that, klass) {
  let set = (key) => {
    thrown new Error(`cannot set "${key}" before render`);
  };

  that.set = (key, value) => { set(key, value) };

  return class extends klass {
    constructor(...args) {
      super(...args);
      set = (key, value) => { Reflect.set(this, key, value) }
    }
  };
}
```

**Comment by chancancode** [comment]
...on second though the above won't exactly work because there are probably existing code that calls `this.set` before render...

So either in the constructor you would have to enumerate all the set properties and apply it on to the component.

Alternatively:

1. Introduce `const state = new TrackedMap()` at the top of each test function
2. Update all `this.set(key, value)` to `state.set(key, value)`

But even then, it still won't work that great, because existing code could (and do) perform `this.set(...)` in `beforeEach` and stuff.

**Pull request closed without merging**